content/posts/2018-10-21-reactive-state.md

```diff
@@ -0,0 +1,353 @@
+A tree can be represented by a state and each drop of rain mutates it,
+increasing the water supply level and applying the pressure on leaves.
+
+The concept is not new, but sometimes it becomes too hard to manage it.
+Even in software development, which was basically created to represent
+the world around us in strict terms.
+
+# Real Life Example
+
+There is a brand-new project which will provide book recommendations.
+The very first step is getting a list of books from a backend.
+This is enough _forever and ever_. Sounds good.
+
+
+```kotlin
+interface BookService {
+    fun getBooks(): Single<List<Book>>
+}
+```
+
+Suddenly — _because Agile_ — we need to save books on the backend.
+All right.
+
+```kotlin
+interface BookService {
+    fun getBooks(): Single<List<Book>>
+    fun createBook(book: Book): Single<Unit>
+}
+```
+
+New features, new screens! Unfortunately, it means that
+the books fetching progress should be preserved across screens.
+A property should do the trick...
+
+```kotlin
+interface BookService {
+    fun getBooks(): Single<List<Book>>
+    val getBooksProgress: Observable<Boolean>
+
+    fun createBook(book: Book): Single<Unit>
+}
+```
+
+Damn, the QA team brought up an issue at the very last minute before the release.
+Fetching books might fail, and we need to show it on all screens
+to give the ability to re-fetch them. Just a sec, another property
+and here we go.
+
+```kotlin
+interface BookService {
+    fun getBooks(): Single<List<Book>>
+    val getBooksProgress: Observable<Boolean>
+    val getBooksFailure: Observable<Boolean>
+
+    fun createBook(book: Book): Single<Unit>
+}
+```
+
+The project hit the production! It works all right, but
+the very first customer had a complaint that the book was created wrong
+and there is no way to delete it. Sounds like creating a book,
+but [some would say it is the reverse](https://www.youtube.com/watch?v=2YTLtG4LMsM)...
+
+```kotlin
+interface BookService {
+    fun getBooks(): Single<List<Book>>
+    val getBooksProgress: Observable<Boolean>
+    val getBooksFailure: Observable<Boolean>
+
+    fun createBook(book: Book): Single<Unit>
+    fun deleteBook(book: Book): Single<Unit>
+}
+```
+
+And then someone brings up that the `BookService` should cache books...
+
+# This is Bad
+
+The `BookService` is far from perfect.
+
+`BookService` clients gradually became more and more complicated.
+Instead of a comfy stateless life they are forced to remember
+that creating a book should trigger re-fetching books from a backend.
+At the same time, this refresh operation should be done only
+on `BookCreateResult.Success` and not on `BookCreateResult.Failure`.
+The same goes to the delete operation.
+Most likely this logic will be distributed and copy-pasted across the client code.
+
+Another distinct feature is how easily the `BookService` transformed from being
+stateless to be stateful. Essentially a pure `getBooks` produced
```

Comment:
hm, it's not technically pure because it does I/O, up2u though

Reply:
Well, it was meant in a broader sense, not being attached to real-life implementation details.